lib/services/role.go

```diff
@@ -1829,45 +1829,56 @@ func (m *DatabaseNameMatcher) String() string {
 }
 
 type loginMatcher struct {
-	login string
+	login       string
+	entitlement string
 }
 
 // NewLoginMatcher creates a RoleMatcher that checks whether the role's logins
 // match the specified condition.
-func NewLoginMatcher(login string) RoleMatcher {
-	return &loginMatcher{login: login}
+func NewLoginMatcher(idemeumId string, login string) RoleMatcher {
+	if idemeumId == "" {
+		return &loginMatcher{login: login}
+	}
+	return &loginMatcher{login: login, entitlement: idemeumId + ":" + login}
 }
 
 // Match matches a login against a role.
 func (l *loginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
-	logins := role.GetLogins(typ)
-	for _, login := range logins {
-		if l.login == login {
-			return true, nil
+	if l.entitlement == "" {
+		return matches(l.login, role.GetLogins(typ)), nil
+	}
+	return matches(l.login, role.GetLogins(typ)) && matches(l.entitlement, role.GetIdemeumEntitlements()), nil
+}
+
+func matches(target string, values []string) bool {
+	for _, value := range values {
+		if target == value {
+			return true
 		}
 	}
-	return false, nil
+	return false
 }
 
 type windowsLoginMatcher struct {
-	login string
+	login       string
+	entitlement string
 }
 
 // NewWindowsLoginMatcher creates a RoleMatcher that checks whether the role's
 // Windows desktop logins match the specified condition.
-func NewWindowsLoginMatcher(login string) RoleMatcher {
-	return &windowsLoginMatcher{login: login}
+func NewWindowsLoginMatcher(idemeumId string, login string) RoleMatcher {
+	if idemeumId == "" {
+		return &windowsLoginMatcher{login: login}
+	}
+	return &windowsLoginMatcher{login: login, entitlement: idemeumId + ":" + login}
 }
 
 // Match matches a Windows Desktop login against a role.
 func (l *windowsLoginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
-	logins := role.GetWindowsLogins(typ)
-	for _, login := range logins {
-		if l.login == login {
-			return true, nil
-		}
+	if l.entitlement == "" {
+		return matches(l.login, role.GetWindowsLogins(typ)), nil
 	}
-	return false, nil
+	return matches(l.login, role.GetWindowsLogins(typ)) && matches(l.entitlement, role.GetIdemeumEntitlements()), nil
```

Comment:
I don't think we can resolve the login names here, we are adding the user traits which is app_id+":"+names
You can only match against the idemeum entitlements here.

Reply:
These login names come from traits, no resolution is happening.